FPAD-7970: Add design doc for persisting interventions (#936)
govuk-one-login/account-interventions-service: docs/design-docs/persisting-interventions.md

# Persisting Interventions

## Introduction

AIS currently maintains account state in a single DynamoDB record per user (the `AccountStatusTable`).
That record is a mutable snapshot of account state: when a new intervention arrives, the previous state is overwritten.
Each account has a JSON object storing the states it's moved through and some data about the relevant interventions, but no data about the states of the interventions themselvesn

> There is no durable log of what interventions were applied, when, or by whom.

Reviewer: Is this true?  I think the existing `history` includes some of this information.

The current history is not kept forever, although this isn't different to this proposal.

Author: this is definitely not true, and I even half rewrote this section to fix this already 🤦

Without a more complete intervention history we cannot:

- Support multiple interventions simultaneously on a single account without having `~O(n^2)` specific states for `n` interventions.
- Add new interventions without adding `n` new states as above, and having a model for how that new intervention is applied.

> - Audit which interventions were applied to an account and in what order.

Reviewer: I don't know if this is true

Author: yeah this is probably not true

- Report on intervention volumes, durations, and outcomes across the estate.
- Support event-driven consumers — if AIS ever does need to act as a smart event bus, downstream services need a reliable ordered log of interventions to subscribe to.

> 

Reviewer: ```suggestion
- Record if an intervention was removed due to it being `superseded`, `mitigated`, or `removed`.
```

> This document proposes persisting interventions immutably in new DynamoDB, written by the existing `InterventionsProcessorFunction` at the point a state transition is successfully applied.

Reviewer: ```suggestion
This document proposes persisting interventions immutably in a new DynamoDB table, written by the existing `InterventionsProcessorFunction` at the point a state transition is successfully applied.
```

> This document proposes persisting interventions immutably in new DynamoDB, written by the existing `InterventionsProcessorFunction` at the point a state transition is successfully applied.

Reviewer: We could also:

1. Enrich the existing `history` field on the existing table.
2. Store a different type of row in the existing table (this is how FAI works).

> This document proposes persisting interventions immutably in new DynamoDB, written by the existing `InterventionsProcessorFunction` at the point a state transition is successfully applied.

Reviewer: Does the new table need to be immutable?

Author: nope, very much the opposite. Will fix

| `createdAt`              | Number        | Sort key      | Unix timestamp in **milliseconds** at which AIS wrote this record. Enables chronological ordering within a partition and supports range queries (e.g. "all interventions for account X since time T"). |
| `interventionId`         | String        | —             | A generated UUIDv4. Provides a stable, globally unique identifier for this intervention record that can be referenced in audit trails, support tickets, and downstream systems without exposing the `accountId`. |
| `interventionCode`       | String        | —             | The numeric code from `Codes` enum (e.g. `"01"`, `"02"`). Preserved verbatim from the ingress event. |
| `interventionName`       | String        | —             | The resolved `AISInterventionTypes` enum value (e.g. `AIS_ACCOUNT_SUSPENDED`). Denormalised here so consumers do not need to join against config. |
| `interventionReason`     | String        | —             | Free-text reason from `extensions.intervention.intervention_reason` in the ingress event. |

> | `interventionState`      | String        | —             | The current state of the intervention: `created`, `applied`, `superseded`, `mitigated`, or `removed`. |

Reviewer: These don't match the statuses from [here](https://docs.google.com/presentation/d/1suP2v-YvJOY2q3X2957hAHIAzqkj6Z4lln5HhE9Isqo/edit?slide=id.g36f16a77f24_0_279#slide=id.g36f16a77f24_0_279)
Should we store `Ignored`?

Author: hm, ignored is an interesting case. I was using the simplified flow chart from the DRB check and forgot about it. I don't think we can answer the question about ignored until the ownership question is sorted, but I suppose in either case we probably store it for reporting reasons:

- if we politely ask other services to apply interventions, then it starts as `created` then as the first step moves to `applied` or `ignored` 
- if we rudely tell other services to apply interventions, then we just write `applied` or `ignored` straight up.

Adding the `applied` status in between `created` and the end states is I think becoming necessary, although I don't have an ironclad argument for it. One to discuss with Chris.

| `sentAt`                 | Number        | —             | Unix timestamp in milliseconds from `event_timestamp_ms` (or `timestamp * 1000`). When the originating system sent the event. |
| `componentId`            | String        | —             | `component_id` from the ingress event. The system that emitted the TxMA event. |
| `originatingComponentId` | String        | —             | `extensions.intervention.originating_component_id`. Optional — not all events carry it. Test fixtures consistently show `component_id: 'TICF_CRI'` paired with `originating_component_id: 'CMS'`, confirming the two fields are not redundant. Both are preserved verbatim; the precise semantic distinction is owned by the upstream systems. |
| `requesterId`            | String        | —             | `extensions.intervention.requester_id`. The identity of the human or system that requested the intervention - omitted for automated interventions. |
| `originatorReferenceId`  | String        | —             | `extensions.intervention.originator_reference_id`. The reference ID in the originating system. |

> | `ttl`                    | Number        | —             | Unix timestamp in **seconds** at which DynamoDB should expire this item. Set to `createdAt / 1000 + retentionSeconds`. Retention period should match the programme's data retention policy (currently 2 years / 63,072,000 seconds, consistent with `DELETED_ACCOUNT_RETENTION_SECONDS`). |

Reviewer: I think we want some sort of history or timestamps, specifically `removedAt`.

Author: Can you be more specific? do you mean when the row gets removed for TTL reasons?

##### Why `accountId` as partition key?

> The dominant access pattern is "give me all interventions for this account".

Reviewer: The other main patterns are "give me all active interventions" and maybe "give me the highest priority intervention".

Author: do you mean all active interventions for a given account? I don't think `highest priority intervention` is something dynamo can provide as part of this design.

An individual account should never have many events, and reading all rows in a given partition is fast in DynamoDB, so it's unlikely performance will become an issue.

---

### Infrastructure

In either case we need a new DynamoDB table called `InterventionStatus`, consistent with the existing `AccountStatusTable`.
See [Appendix I](#appendix-i---dynamodb-cloudformation-yaml) for suggested `yaml`.

A dedicated KMS key (`InterventionStatusTableEncryptionKey`) with the same key policy pattern as `AccountStatusTableEncryptionKey` must be created alongside the table.
The table ARN and encryption key ARN must be exported as CloudFormation outputs so the `ais-main` stack can import them.
`InterventionsProcessorFunctionManagedPolicy` must be extended with `dynamodb:PutItem` on the new table ARN and `kms:Decrypt` / `kms:GenerateDataKey` on the new table's KMS key.
No other Lambda roles require access to this table — it is write-only from the application's perspective.

The option exists of using a dedicated queue and Lambda to decouple the write from the main processing path.
This is probably unnecessary:

> - `InterventionsProcessorFunction` already performs a `dynamodb:UpdateItem` on `AccountStatusTable` synchronously in the hot path. Adding a second `dynamodb:PutItem` is the same class of operation with the same latency profile. It's difficult to find solid numbers for send latency on SQS but it's unlikely to be a critical difference.

Reviewer: This is potentially more complicated for option A, as we may need to update existing interventions we will need to `Query` the existing interventions and then potentially `Put` or `Update` multiple interventions.

Author: yes, this is true

Author: this is a hot mess actually - we need to grab all interventions history in case there's something we need to supersede right?

- If the write fails, the existing retry mechanism (SQS `ReportBatchItemFailures`) will re-drive the message, so the history record will eventually be written when the state transition is retried.

> It may be worth considering using [DynamoDB transactions](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/transactions.html) to avoid the case where we only write successfully to either the account table or the interventions table.

Reviewer: Depending on how complicated the new logic is, there may be an attraction to decoupling these, so that we aren't blocking `AccountStatusTable` updates if our new logic fails.

Author: I think that's a big question - if our new table isn't a first-class source of truth, it's pretty awkward to be able to trust it to do anything. but if it is, we either need transactions or we need to cascade the operations in a sensible way

Some quick googling implies this has performance impacts.
An alternative is making the writes idempotent - this requires a read for every write so is probably not feasible.

---

### Interfaces

We aim to have all persisted data strictly parsed at write time. 

> Because this data is entirely under our control (i.e. it's not part of the event catalogue and the schema isn't consumed by other components of DI) we can do this using Zod.

Reviewer: Potentially, if we are strictly validating our inputs, it may not be necessary to revalidate before writing to the database.

Author: We might be talking at cross-purposes but the storage service doesn't know anything about inputs, aside from its own interface. Having a data-drive parsing layer before the write is in some ways about convenience - it gives us confidence we're getting the same things in and out, and that confidence is derived from data (Zod objects) instead of from code (the storage service's code)

We should consider taking the opportunity to refactor how we communicate with DynamoDB to be re-usable and more in line with a ports-and-adapters approach, for increased flexibility and testability.

Because of DynamoDB's lack of schema enforcement and the possibility of the data schema needing to evolve over time, we should consider versioning each row in the table and storing that version in the row itself.

> This allows us to parse based on version at write-time and read-time, and easily alter handling between versions and handle deprecation scenarios etc.

Reviewer: At some point we should think a bit about how we are going to migrate data to this new world.  Presumably to start with we aren't going to assume any records exist in this new table, and drive entirely from the status in `AccountStatusTable`.

Author: Yeah, 100%. For now the only thing in scope is storing the things - at some point we'll have to move to a hybrid model where we check both tables, and hopefully eventually move to a model where we can trust just the intervention storage. I haven't thought through whether the account table is still worth having in that state

##### DeletionProtection
Must be enabled in all non-dev environments, consistent with `AccountStatusTable`.

##### TTL-based expiry
Items expire automatically via DynamoDB TTL.
This is a soft delete — DynamoDB removes expired items asynchronously within 48 hours.
The application must not rely on TTL for hard deletion guarantees; if regulatory requirements demand hard deletion within a specific window, a separate cleanup process is needed.

##### Write failure handling
A failed `PutItem` causes the SQS record to be returned as a `batchItemFailure`.
The message is retried up to `maxReceiveCount` (currently 60) times before being sent to the DLQ.
The DLQ alarm (`InterventionsProcessorErrorsCanary`) will fire if errors persist.
This is the same resilience model as the existing state write and requires no additional infrastructure.

##### No single point of failure
DynamoDB is a managed, multi-AZ service.
There is no additional availability concern introduced by this change.

### Performance and Scalability

##### Latency budget
`InterventionsProcessorFunction` has a 5-second timeout.
The existing processing path includes one `Query` and one `UpdateItem` against `AccountStatusTable`.

> Adding one `PutItem` against the history table adds approximately 2–5 ms at p99 (same-region DynamoDB).

Reviewer: For Option A there are more operations required.

Author: yeah, this needs much more thought

      SSEType: KMS
    DeletionProtectionEnabled: !If [IsDevEnvironment, false, true]
    PointInTimeRecoverySpecification:
      PointInTimeRecoveryEnabled: true
    TimeToLiveSpecification:
      AttributeName: ttl
      Enabled: true
    Tags:
      - Key: BackupFrequency
        Value: Bihourly
      # ... standard programme tags
```

### Appendix ii - Typescript interfaces

#### Option A interfaces

##### TypeScript type: `InterventionHistoryRecord`

```typescript
export type InterventionHistoryRecord {
  accountId: string;
  createdAt: number;           // ms epoch, set by AIS at write time
  interventionId: string;      // UUIDv4
  interventionCode: string;    // Codes enum value
  interventionName: string;    // AISInterventionTypes enum value
  interventionReason: string;
  interventionState: string;   // created | applied | superseded | mitigated | removed
  sentAt: number;              // ms epoch from ingress event
  componentId: string;
  originatingComponentId?: string;
  requesterId?: string;
  originatorReferenceId?: string;
  ttl: number;                 // seconds epoch
}
```

##### Service interface

```typescript
export interface InterventionStatusService {

>   persistIntervention(record: InterventionHistoryRecord): Promise<void>;

Reviewer: Will the logic inside this work out if an existing intervention exists and update or create accordingly?

Author: Probably not worth answering that question - this needs a lot of expanding 😅

>   persistIntervention(record: InterventionHistoryRecord): Promise<void>;

Reviewer: How will this work when `superseding` an intervention?  Will lots of the fields be ignored, or will those fields be overwritten?

Author: see above 😅

Each account has a JSON object storing the states it's moved through and some data about the relevant interventions, but no data about the states of the interventions themselvesn
Without a more complete intervention history we cannot:

- Support multiple interventions simultaneously on a single account without having `~O(n^2)` specific states for `n` interventions.
- Add new interventions without adding `n` new states as above, and having a model for how that new intervention is applied.
- Report on intervention volumes, durations, and outcomes across the estate.
- Support event-driven consumers — if AIS ever does need to act as a smart event bus, downstream services need a reliable ordered log of interventions to subscribe to.
- Record if an intervention was removed due to it being `superseded`, `mitigated`, or `removed`.

This document proposes persisting interventions in a new DynamoDB table, written by the existing `InterventionsProcessorFunction` at the point a state transition is successfully applied.

## Assumptions

- Interventions are independent between accounts

> - DynamoDB means we don't have to worry about scaling

Reviewer: I think this is a given and can be removed

```suggestion
```

> 

Reviewer: ```suggestion
- The number of interventions per account is not so large that we can't fetch all the interventions and handle them in code.
```

This means a natural option is to store all rows in the same table and tag them so we know what the difference is.
Consequences of this co-mingling approach compared with an approach using multiple tables are:

- We don't need to create or maintain another table and the access to that table
- We need to tag rows so we know what kind of object they represent
- We need to handle existing rows which aren't tagged
- We need to use the existing sort key on the account status table to do queries on interventions
- We need to add logic to the application to handle the different kinds of rows and tags
- We wouldn't be trivially able to throw all the intervention rows away and start again

Based on these consequences, using a co-mingling approach would necessitate extra complexity in the application and the table itself while saving the cost of creating a new table.
It also removes our ability to trivially throw away the intervention rows while the table is write-only, which may be a useful step while iterating on the table specification.
Setting up a new DynamoDB table is cheap and fairly simple, and we have existing infrastructure-as-code that we can use by analogy.

> Because of this, we're choosing a multiple table approach for this problem, and **not** using a co-mingling approach.

Reviewer: I think we are agreed on this, can we move it to the end as another option considered and rejected?

##### Why `accountId` as partition key?
The dominant access pattern is "give me all interventions for this account".
A partition key of `accountId` makes that a single `Query` call with no scan.
Secondary access patterns (e.g. reporting across all accounts) should be served by an export pipeline, not by querying this table directly.

##### Why `createdAt` (write time) as sort key?
`createdAt` is set by AIS at write time and is always increasing within a partition, giving a reliable ordering of_when AIS processed each intervention.
`sentAt` is controlled by the upstream system and is not guaranteed to be monotonically increasing.
Events can arrive out of order or be replayed, in which case AIS might already have taken action before the "first" of two events arrives.
This means for ordering purposes we should rely on when AIS receives an event, which most of the time will provide the same ordering as `sentAt`
`sentAt` is preserved as a separate attribute for audit purposes.

---

#### Option B: Multiple immutable events per-intervention (event sourcing)

> Rather than storing one row per intervention, this approach stores one row per event that happens to an intervention.

Reviewer: I presume this means that one "transaction" can produce multiple rows, for example a permanent suspension might undo a temporary suspension?

The current state of any intervention is derived by replaying all events for a given `accountId` in order.
No rows are ever updated or deleted (except by TTL expiry).

An intervention moves through a state machine:

```
created → applied → superseded
                  → mitigated
                  → removed
```

> `applied` is a distinct state from `created` because application of an intervention may be asynchronous — the intervention can be recorded as `created` before AIS has confirmed it has taken effect.

Reviewer: I think we would want to store a row for `ignored` events as well

##### Least-privilege IAM
The `InterventionsProcessorFunction` role must be granted only `dynamodb:PutItem` on the history table.
No Lambda role should be granted `dynamodb:Query`, `dynamodb:Scan`, or `dynamodb:GetItem` on this table — it is write-only from the application.
Read access for audit and reporting purposes must be granted separately to dedicated roles outside the application stack.

##### Data classification
`accountId` is a sensitive identifier.
Log statements that include `accountId` must be prefixed with `LOGS_PREFIX_SENSITIVE_INFO` (consistent with existing practice) and filtered from CSLS subscription filters.

### Resilience

##### Point-in-time recovery (PITR)
Must be enabled.
PITR provides a continuous backup with a 35-day recovery window at no additional operational overhead.
This is already enabled on `AccountStatusTable` and must be consistent here.

##### DeletionProtection
Must be enabled in all non-dev environments, consistent with `AccountStatusTable`.

##### TTL-based expiry
Items expire automatically via DynamoDB TTL.
This is a soft delete — DynamoDB removes expired items asynchronously within 48 hours.

> The application must not rely on TTL for hard deletion guarantees; if regulatory requirements demand hard deletion within a specific window, a separate cleanup process is needed.

Reviewer: At some point we need to specify how this works, I think for now "the same as the current `history`" will suffice

### Observability

##### Metrics
We should record a counter for writes to the new table(s) which records successes and errors so we can see throughput and error rate.
We should consider recording the status and intervention type within the metric labels as well to make it easy to get basic metrics for the service.

We should log thrown errors, so we can debug them - these must contain no sensitive information.

### Testability

##### Ports-and-adapters pattern
The persistence interface decouples the handler from the DynamoDB implementation.
Unit tests can then inject a fake implementation.

##### Integration tests
The existing feature tests in `feature-tests/` use a real DynamoDB table.
The relevant infrastructure will be deployed identically across environments making the feature test behaviour identical.

##### Mutation testing
The new service is in scope for Stryker.
The `persistIntervention` method and its callers should achieve the same mutation score threshold as the rest of the codebase.

## Rejected Solutions

### Co-mingling data in the table

> Every case we discuss in this document involves using account ID as the partition key, and the current account status table also uses account ID as a partition key.
> This means a natural option is to store all rows in the same table and tag them so we know what the difference is.
> Consequences of this co-mingling approach compared with an approach using multiple tables are:
> 
> - We don't need to create or maintain another table and the access to that table
> - We need to tag rows so we know what kind of object they represent
> - We need to handle existing rows which aren't tagged
> - We need to use the existing sort key on the account status table to do queries on interventions
> - We need to add logic to the application to handle the different kinds of rows and tags
> - We wouldn't be trivially able to throw all the intervention rows away and start again
> 
> Based on these consequences, using a co-mingling approach would necessitate extra complexity in the application and the table itself while saving the cost of creating a new table.
> It also removes our ability to trivially throw away the intervention rows while the table is write-only, which may be a useful step while iterating on the table specification.
> Setting up a new DynamoDB table is cheap and fairly simple, and we have existing infrastructure-as-code that we can use by analogy.
> Because of this, we're choosing a multiple table approach for this problem, and **not** using a co-mingling approach.

Reviewer 2: (a reminder about the outstanding edit to this section)

```suggestion
Every option described here uses the account ID as the partition key. Currently the account status table also uses the same partition key. This means we could store the proposed data in the existing table. 

Whilst avoiding initial overhead in creating a new table, we prefer to create a new table for this data. This preference is because co-mingling would:

- introduce prefixing values of key(s), managing legacy records, and adding complex application logic to distinguish record types – work which would have little immediate value 
- inhibit throwing away the new table and start again, if the need occurred – something which may be valuable during early iteration of the schema design whist treat the data as write only

The approach now for using a new discrete table does not preclude co-mining data in the future. 
```
